controllers/topology/desired_state.go

```diff
@@ -301,6 +371,93 @@ func computeMachineDeployment(_ context.Context, s *scope.Scope, machineDeployme
+		return currentVersion, nil
+	}
+
+	// Return early if the currentVersion is already equal to the desiredVersion
+	// no further checks required.
+	if currentVersion == desiredVersion {
+		return currentVersion, nil
+	}
+
+	// If the control plane is being created (current control plane is nil), do not perform any machine deployment upgrade in this case.
+	// Return the current version of the machine deployment.
+	// NOTE: this case should never happen (upgrading a MachineDeployment) before creating a CP, but we are implementing this check for extra safety.
+	if s.Current.ControlPlane == nil || s.Current.ControlPlane.Object == nil {
+		return currentVersion, nil
+	}
+
+	// If the current control plane is upgrading, then do not pick up the desiredVersion yet.
+	// Return the current version of the machine deployment. We will pickup the new version after the control
+	// plane is stable.
+	cpUpgrading, err := contract.ControlPlane().IsUpgrading(s.Current.ControlPlane.Object)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to check if control plane is upgrading")
+	}
+	if cpUpgrading {
+		return currentVersion, nil
+	}
+
+	// If control plane supports replicas, check if the control plane is in the middle of a scale operation.
+	// If the current control plane is scaling, then do not pick up the desiredVersion yet.
+	// Return the current version of the machine deployment. We will pickup the new version after the control
+	// plane is stable.
+	if s.Blueprint.Topology.ControlPlane.Replicas != nil {
+		cpScaling, err := contract.ControlPlane().IsScaling(s.Current.ControlPlane.Object)
+		if err != nil {
+			return "", errors.Wrap(err, "failed to check if the control plane is scaling")
+		}
+		if cpScaling {
+			return currentVersion, nil
+		}
+	}
+
+	// Check if we are about to upgrade the control plane. In that case, do not upgrade the machine deployment yet.
+	// Wait for the new upgrade operation on the control plane to finish before picking up the new version for the
+	// machine deployment.
+	currentCPVersion, err := contract.ControlPlane().Version().Get(s.Current.ControlPlane.Object)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get version of current control plane")
+	}
+	computedCPVersion, err := contract.ControlPlane().Version().Get(desiredControlPlaneState.Object)
+	if err != nil {
+		return "", errors.Wrap(err, "failed to get version of desired control plane")
+	}
+	if *currentCPVersion != *computedCPVersion {
+		// The versions of the current and desired control planes do no match, implies we are about to upgrade the control plane.
+		return currentVersion, nil
+	}
+
+	// At this point the control plane is stable (not scaling, not upgrading, not being upgraded).
+	// Checking to see if the machine deployments are also stable.
+	// If any of the MachineDeployments is upgrading, do not upgrade the machine deployment yet.
+	if s.Current.MachineDeployments.AnyIsUpgrading() {
```

Comment:
This assumes only 1 MachineDeployment can be upgraded at the same time while `AllowUpgrade` bases it's decision on the const `maxMachineDeploymentUpgradeConcurrency` (which is *at the moment* also 1). So I think the current behaviour is:

* don't trigger another MD rollout if there's already one in progress
* if there's none in progress: trigger at most 1 new rollout 

I think this behaviour is fine, but I think we have to make the dependency between the thresholds in `AnyIsUpgrading` and `AllowUpgrade` more explicit.

Note: The current behaviour would be a bit inconsistent if we change `maxMachineDeploymentUpgradeConcurrency` to something > 1. As we could have some rollouts in progress before the reconcile and trigger a few more during reconcile. I *don't* think we have to handle this cases right now.

I see the following options:
1. drop maxMachineDeploymentUpgradeConcurrency, hard-code `1` in the check and add a comment about the dependency to AnyIsUpgrading
1. keep AllowUpgrade as is and make it very clear in godoc that maxMachineDeploymentUpgradeConcurrency *must* be 1 or otherwise we need a refactoring
1. use maxMachineDeploymentUpgradeConcurrency consistently for both checks